migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs

```diff
@@ -114,6 +119,82 @@ impl SqlFlavour for PostgresFlavour {
         Ok(())
     }
 
+    async fn apply_migration_script(
+        &self,
+        migration_name: &str,
+        script: &str,
+        conn: &Connection,
+    ) -> ConnectorResult<()> {
+        let (client, _url) = conn.unwrap_postgres();
+        let inner_client = client.client();
+
+        match inner_client.simple_query(script).await {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                let (database_error_code, database_error): (Option<&str>, _) = if let Some(db_error) = err.as_db_error()
+                {
+                    let position = if let Some(ErrorPosition::Original(position)) = db_error.position() {
+                        let mut previous_lines = [""; 5];
+                        let mut byte_index = 0;
+                        let mut error_position = String::new();
+
+                        for (line_idx, line) in script.lines().enumerate() {
+                            // Line numbers start at 1, not 0.
+                            let line_number = line_idx + 1;
+                            byte_index += line.len() + 1; // + 1 for the \n character.
+
+                            if *position as usize <= byte_index {
+                                error_position = format!(
+                                    "\n\nPosition:\n{}\n\x1b[1m{:>3}\x1b[1;31m {}\x1b[0m",
+                                    previous_lines
+                                        .iter()
+                                        .enumerate()
+                                        .filter_map(|(idx, line)| line_number
+                                            .checked_sub(previous_lines.len() - idx)
+                                            .map(|idx| (idx, line)))
+                                        .map(|(idx, line)| format!(
+                                            "\x1b[1m{:>3}\x1b[0m{}{}",
+                                            idx,
+                                            if line.is_empty() { "" } else { " " },
+                                            line
+                                        ))
+                                        .join("\n"),
+                                    line_number,
+                                    line
+                                );
+                                break;
+                            } else {
+                                previous_lines = [
+                                    previous_lines[1],
+                                    previous_lines[2],
+                                    previous_lines[3],
+                                    previous_lines[4],
+                                    line,
+                                ];
+                            }
+                        }
```

Comment:
This should probably move into its own file.

Reply:
What's going on in here? What should this code be doing?

Reply:
It prints the line where the error happened in red, with surrounding context and line numbers.